### PR TITLE
build from source on windows: build_all.bat

### PR DESCRIPTION
**Pull request by timotheecour**
## SCRATCH below

EDIT1: from IRC/gitter chat
Varriount @Varriount Aug 10 22:14
@timotheecour https://gist.github.com/Varriount/8f8c0d5e5224458cbf1dcf1a8916473e
Note, that's an untested prototype.

timothee:
* there’s a bug in your code, you skipped cd csources so it pulls from nim, not csources
* why no || goto:error after koch tools?
* for the last paragraph (debug mode), isn’t that what ./koch temp is for ?

Timothee Cour @timotheecour Aug 10 23:58
* btw, the path issue is still annoying for me; I have to resort to a kludge like:
setx path “C:\… C:\Users\User\timw\nim-0.18.0_x64\nim-0.18.0\dist\mingw64\bin\” which is ugly, because finish.exe didn’t work for me (to get gcc in path); any idea?
* how about using build_all.py instead of build_all.bat ???? if user wants to hack on nim soures, asking him to have python installed is OK IMO

Varriount @Varriount Aug 11 04:32
@timotheecour The thing is, the user may have a 32-bit compiler installed on a 64-bit operating system.

### CODE REVIEW

File: build_all.bat [hunk 1]
```diff
@@ -0,0 +1,44 @@
+echo "build form source on windows"
```
Comment (euantorano) [inline build_all.bat]: Spelling: Should be `build from source`

Comment (timotheecour) [reply build_all.bat]: done

**Comment by timotheecour** [comment]
PTAL

File: build_all.bat [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+rem one way is via calling finish.exe in command prompt after installing via
+rem https://nim-lang.org/install_windows.html
+rem note: this won't work if it's in a networked drive
+rem eg if using parallels on OSX, don't use a parallels shared folder, use C:\
+
+rem better to be verbose, for debugging purposes
+@echo on
+
+echo "checks that gcc in PATH:"
+gcc -v
```
Comment (Varriount) [inline build_all.bat]: The script will continue to run, even if this command fails.

File: build_all.bat [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+IF EXIST csources\nul (
```
Comment (Varriount) [inline build_all.bat]: Why is there a null escape here?

File: build_all.bat [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  echo "csources already exists"
+) ELSE (
+  git clone --depth 1 https://github.com/nim-lang/csources.git
+)
+
+rem nesting this inside an IF branch triggered a weird parsing error
+rem https://stackoverflow.com/questions/601089/detect-whether-current-windows-version-is-32-bit-or-64-bit
+if defined ProgramFiles(x86) (
```
Comment (Varriount) [inline build_all.bat]: See https://stackoverflow.com/questions/12322308/batch-file-to-check-64bit-or-32bit-os

File: build_all.bat (hunk 2 above)
Comment (timotheecour) [reply build_all.bat]: fixed using ERRORLEVEL everywhere relevant

File: build_all.bat (hunk 3 above)
Comment (timotheecour) [reply build_all.bat]: fixed (that depended on the stackoverflow answers)

File: build_all.bat (hunk 4 above)
Comment (timotheecour) [reply build_all.bat]: I tried, it didnt' work on command prompt

**Comment by timotheecour** [comment]
PTAL all comments addressed

**Comment by dom96** [comment]
Update readme too?

File: build_all.bat [hunk 5]
```diff
@@ -0,0 +1,76 @@
+rem "build from source on windows"
+
+@echo off
+echo requires `gcc` in your PATH
```
Comment (dom96) [inline build_all.bat]: I'm not familiar with batch, but does this get echoed no matter what?

Wouldn't it be better to echo it only if ``gcc`` isn't in the PATH?

File: build_all.bat [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  git clone --depth 1 https://github.com/nim-lang/csources.git
+  IF %ERRORLEVEL% neq 0 (GOTO:END)
+)
+
+rem this didn't work in command prompt:
+rem https://stackoverflow.com/questions/12322308/batch-file-to-check-64bit-or-32bit-os
+rem reg Query "HKLM\Hardware\Description\System\CentralProcessor\0" | find /i "x86" > NUL && set OS=32BIT || set OS=64BIT
+
+rem https://stackoverflow.com/questions/601089/detect-whether-current-windows-version-is-32-bit-or-64-bit
+if defined ProgramFiles(x86) (
```
Comment (dom96) [inline build_all.bat]: You should detect whether the ``gcc`` is 32bit or 64bit instead. Otherwise this will fail on Win64 with a 32 bit gcc.

**Comment by Araq** [comment]
I don't think we want to maintain this, Windows users have it rather easy with the prebuilt binaries anyway. Closing.

**Comment by dom96** [comment]
So we're happy with build_all.sh but not build_all.bat?

**Comment by Araq** [comment]
Yeah, why not.

**Comment by timotheecour** [comment]
> I don't think we want to maintain this, Windows users have it rather easy with the prebuilt binaries anyway. Closing.

the reason I even did this PR was because it's such a pain to build from source on windows. Prebuilt binaries doesn't help when we need to build from source.
Example use case:
* when writing a PR, and tests fail only on windows, and we need to test things locally.
* when developping compiler on windows

Note: I dislike windows batch files as much as anyone else, but hiding things under a rug doesn't make anything easier.
Would you guys be ok with a **python based install script for windows**? (would probably be more robust and easier to maintain/ read; only drawback is a dependency on python but not a big deal if you're assuming a developper is already willing to compile nim from source)

**Comment by Araq** [comment]
> Would you guys be ok with a python based install script for windows

No. 

> the reason I even did this PR was because it's such a pain to build from source on windows. Prebuilt binaries doesn't help when we need to build from source.

Yeah, they do as an old koch, nim can usually be used to build a new Nim.

> Example use case

Plenty of people survive with building from source on Windows. Just like building from source on Unix was fine before. One is simply better off understanding the build process and one can always keep a personal script around. Not to mention ``choosenim`` can also give you a development version.

**Comment by dom96** [comment]
Honestly, if you're developing Nim/writing PRs you should know how to build Nim (it's not difficult).

Now our instructions are interspersed with ``build_all.sh`` and the custom Windows instructions in the readme, the way it was before was IMO better. Developers should understand how bootstrapping works instead of running scripts which do god knows what.

**Pull request closed without merging**